test/functional/p2p_blocksonly.py

```diff
@@ -69,18 +49,55 @@ def run_test(self):
         peer_2_info = self.nodes[0].getpeerinfo()[1]
         assert_equal(peer_2_info['whitelisted'], True)
         assert_equal(peer_2_info['permissions'], ['noban', 'forcerelay', 'relay', 'mempool', 'download'])
-        assert_equal(self.nodes[0].testmempoolaccept([sigtx])[0]['allowed'], True)
-        txid = self.nodes[0].testmempoolaccept([sigtx])[0]['txid']
+        assert_equal(self.nodes[0].testmempoolaccept([tx_hex])[0]['allowed'], True)
 
         self.log.info('Check that the tx from forcerelay first_peer is relayed to others (ie.second_peer)')
         with self.nodes[0].assert_debug_log(["received getdata"]):
-            first_peer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+            first_peer.send_message(msg_tx(tx))
             self.log.info('Check that the forcerelay peer is still connected after sending the transaction')
             assert_equal(first_peer.is_connected, True)
             second_peer.wait_for_tx(txid)
             assert_equal(self.nodes[0].getmempoolinfo()['size'], 1)
         self.log.info("Forcerelay peer's transaction is accepted and relayed")
 
+        self.nodes[0].disconnect_p2ps()
+        self.nodes[0].generate(1)
+
+    def blocks_relay_conn_tests(self):
+        self.log.info('Tests with node in normal mode with block-relay-only connections')
+        self.restart_node(0, ["-noblocksonly"])  # disables blocks only mode
+        assert_equal(self.nodes[0].getnetworkinfo()['localrelay'], True)
+
+        # Ensure we disconnect if a block-relay-only connection sends us a transaction
+        self.nodes[0].add_outbound_p2p_connection(P2PInterface(), connection_type="blockrelay")
+        assert_equal(self.nodes[0].getpeerinfo()[0]['relaytxes'], False)
+        _, txid, tx_hex = self.check_p2p_tx_violation(index=2)
+
+        self.log.info("Check that txs from RPC are not sent to blockrelay connection")
+        conn = self.nodes[0].add_outbound_p2p_connection(P2PTxInvStore(), connection_type="blockrelay")
+
+        assert_equal(self.nodes[0].testmempoolaccept([tx_hex])[0]['allowed'], True)
+        self.nodes[0].sendrawtransaction(tx_hex)
+        conn.sync_with_ping()
+        assert(txid not in conn.get_invs())
+
+    def check_p2p_tx_violation(self, index=1):
+        self.log.info('Check that txs from P2P are rejected and result in disconnect')
+        input_txid = self.nodes[0].getblock(self.nodes[0].getblockhash(index), 2)['tx'][0]['txid']
+        tx = create_transaction(self.nodes[0], input_txid, self.nodes[0].getnewaddress(), amount=50 - 0.001)
```

Comment:
`getnewaddress` requires a wallet, so this test fails without the wallet compiled. Can you replace it with a hard-coded address?

Reply:
so simple! done :)

Reply:
ah, looks like the blocktools helper uses wallet, so I ended up just adding `skip_if_no_wallet` and reverting this.

Reply:
ok, it's a bit of a shame to require the wallet to be compiled to run this test. Ideally, I think we'd want just the `wallet_` tests to require the wallet, but we're a long way from that.

No need to change this here, but a useful future project would be to eliminate the wallet dependency from the non-wallet tests.

Reply:
I haven't looked too closely at #19800 yet, but I think I should now be able to replace this `create_transaction` call with `miniwallet.send_self_transfer` and remove the wallet dependency? does this sound right? cc @MarcoFalke

Reply:
> does this sound right? cc @MarcoFalke

no idea ;)

Just try it and let me know if you run into any problems. If there are bugs in miniwallet, feel free to fix them along the way.

Reply:
OCD nit, maybe enclose `50 - 0.001` in parens, or better yet, assign this magical entity to a well-named variable (or with a comment) that documents what/why

Reply:
resolving this conversation, tracking as a future improvement at https://github.com/bitcoin/bitcoin/pull/19315#issuecomment-735282520